cloudsync/providers/file.py

```diff
@@ -0,0 +1,504 @@
+import os
+import sys
+import glob
+import time
+import re
+import tempfile
+import logging
+import shutil
+import collections
+import threading
+from hashlib import blake2b
+import typing
+from dataclasses import dataclass
+
+from watchdog import events as watchdog_events
+from watchdog.observers import Observer as watchdog_observer
+
+from cloudsync.event import Event
+from cloudsync.provider import Provider
+from cloudsync.registry import register_provider
+from cloudsync.types import OInfo, OType, DirInfo
+import cloudsync.exceptions as ex
+
+
+def is_osx():
+    return sys.platform == "darwin"
+
+
+if is_osx():
+    from Foundation import NSURL  # pylint: disable=import-error,no-name-in-module
+
+
+logging.getLogger("watchdog").setLevel(logging.INFO)
+log = logging.getLogger(__name__)
+
+
+def get_hash(dat):
+    """Returns a hash of it's argument which can be a bytes or filelike object"""
+    # 32-byte blake optimized for 64 bit chips
+    if not hasattr(dat, "read"):
+        return blake2b(dat, digest_size=32).digest()
+    blocksize = 4096
+    d = blake2b(digest_size=32)
+    block = dat.read(blocksize)
+    while block:
+        d.update(block)
+        block = dat.read(blocksize)
+    return d.digest()
+
+
+def detect_case_sensitive(tmpdir=None):
+    """Returns true if temp direcory specified is case sensitive, or gettempdir() if unspecified."""
+    if not tmpdir:
+        tmpdir = tempfile.gettempdir()
+    f1 = os.path.join(tmpdir, "tmp." + os.urandom(16).hex())
+    with open(f1, "w") as f:
+        f.write("x")
+    try:
+        f2 = f1.upper()
+        if os.path.exists(f2):
+            return False
+    finally:
+        os.unlink(f1)
+    return True
+
+
+def casedpath(path):
```

Comment:
should this be called `display_path` or `canonicalize_path`?

Reply:
changed